[wip] kvserver: auto-tune closed timestamp updates

wenyihu6 opened the pull request:
Note that this pr is based on top of https://github.com/cockroachdb/cockroach/pull/141989.

---

Previously, for global reads, we calculated the target closed timestamp using
the `LEAD_FOR_GLOBAL_READS` policy. This policy indicates that the range’s closed
timestamp is configured to lead the present time so that all follower replicas
of the range can serve present time reads. Note that this is an estimate/goal
but not the lead time itself.

This is how the calculation is currently done:

```
raft_propagation_time = max_network_rtt*1.5 (closed timestamp propagation by raft from a leader)
+ raft_overhead(proposal evaluation, log sync, and state machine apply)

side_propagation_time = max_network_rtt*0.5 (closed timestamp propagation by side transport)
+ side_transport_close_interval

// Maximum time required to propagate a closed timestamp update to followers from the sender.
closed_ts_propagation = max(raft_propagation_time, side_propagation_time)

closed_ts_at_sender = now(sender’s current clock) + maxClockOffset (max clock uncertainty)
+ closed_ts_propagation
```

By default, this results in a closed timestamp leading the sender's current clock by 800ms. Users
can override this by tuning `kv.closed_timestamp.lead_for_global_reads_override`.

**Problems with the previous approach**

1. Tuning trade-offs: A higher lead timestamp forces leaseholders to wait longer
for the current time to catch up, incurring a high write latency. A lower lead
timestamp may result in followers not fully replicating the changes before
returning control to the client, forcing follower replicas to redirect query to
the leaseholder and increase read latency.

2. Cluster wide setting: This setting applies to the entire cluster, not per
range. For geographically distant replicas, a short lead time may not be enough
for full application. However, increasing the lead time cluster wide can harm
write performance for replicas that are close together and don’t require as much
time to replicate.

**Improvements in this patch**

This patch auto-tunes closed timestamp propagation for global reads by providing
more accurate estimates for raft and side transport propagation time.

1. Raft Propagation Time Adjustment

Instead of `raft_propagation_time= max_network_rtt*1.5+ raft_overhead`, the
`max_network_rtt*1.5` factor is now replaced with a smoothed average time taken
between a proposal being submitted to raft and local application at leader replica.
This measurement considers only successful write commands, and it does not
include the full application on follower replicas.

Ideally, we would account for full application time on follower replicas, but
the leader has no direct visibility into when follower replicas apply updates to
their state machine. Tracking this propagation across all followers seems
complex, so for now, we rely on raft_overhead to approximate this delay.

2. Side Transport Propagation Time Adjustment

Instead of `side_propagation_time = max_network_rtt*0.5 +
side_transport_close_interval`, the `max_network_rtt*0.5` factor is now replaced
by a smoothed average max latency observed by the sender node. This average
considers the maximum latency across all nodes where the sender has leaseholder
ranges and needs to send closed timestamps to nodes with follower replicas.

Ideally, every node-to-node rpc would have its own network latency average,
implying a different closed timestamp target for each pair. However, this would
complicate the side transport design. Currently, senders and receivers share
state by using a single closed timestamp for each policy, with any new or
removed ranges noted in the group. To assign a unique timestamp for each range
based on different network latencies, we would need to introduce a node-to-node
latency mapping on top of the group ranges state view. In addition, the side
transport propagation calculation is usually dominated by
sideTransportCloseInterval (200ms by default). To avoid the complexity, we
simply use the maximum latency to estimate the one hop network latency.

Resolves: https://github.com/cockroachdb/cockroach/issues/59680
Release note: `kv.closed_timestamp.lead_for_global_reads_auto_tune.enabled` 
can now be used to auto-tune the lead time that ranges global tables use to publish
close timestamps. The `kv.closed_timestamp.lead_for_global_reads_override`
cluster setting takes precedence over this one. If auto-tuning is disabled or
no data is observed, the system falls back to a hardcoded computed lead time.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141832)
<!-- Reviewable:end -->

wenyihu6 commented:
TODO: add unit test, add a cluster setting to guard this, more testing to make sure using the weighted average time idea here is fine

wenyihu6 commented:
Wanted to have some high level feedback and check two things I'm not sure: 

1. Raft propagation estimate: Ideally, we’d account for the full application time on followers, but the leader lacks visibility into when follower replicas apply to their state machine. Tracking this across all followers is complex, so we just approximate the delay using raft_overhead.

2. Side transport estimate: Every node to node RPC should have its own latency average, but this would complicate side transport design. Currently, senders and receivers share a single closed timestamp per policy https://github.com/cockroachdb/cockroach/blob/c83c57d354741ac36740894f7387c014eb6c09fe/pkg/kv/kvserver/closedts/sidetransport/sender.go#L94-L108. Having a different closed  timestamps target would require a node-to-node latency mapping shared between sender and receiver. Since propagation is mostly dominated by sideTransportCloseInterval (200ms), can we simplify by just choosing the maximum observed latency to any follower nodes?

wenyihu6 committed:
kvserver: add kv.closed_timestamp.lead_for_global_reads_auto_tune_interval

This commit adds a new cluster setting,
`kv.closed_timestamp.lead_for_global_reads_auto_tune_interval` to control how
frequently the system measures latency. If non-zero, this observed latency is
used to auto-tune the lead time for ranges using the LEAD_FOR_GLOBAL_READS
policy. The setting defaults to 5 minutes and can be disabled by setting to 0.
It only takes effect when `side_transport_interval` is non-zero.

Note that setting is currently unused, and no release notes have been added yet.
Future commits will populate and integrate it.

Part of: #59680
Release note: none

wenyihu6 commented:
Did some more testing - I found this patch improved things, but the miss rate is still quite high. Going to explore other optimizations on the heuristics here.

wenyihu6 commented:
Converting this pr to draft for now while I explore a better approach to reduce the miss rate.

wenyihu6 committed:
kvserver/closedts: change TargetForPolicy to take network rtt

This patch changes TargetForPolicy to take in network round trip latency.
Currently, it always takes in closedts.DefaultMaxNetworkRTT. But future commits
will change it to take in a dynamically observed network latency.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/closedts: add acceptable rtt range to TargetForPolicy

This patch changes TargetForPolicy to clamp the latency taken in within an
acceptable latency range. Currently, it always takes in
closedts.DefaultMaxNetworkRTT. But future commits will change it to take in a
dynamically observed network latency.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/closedts: replace roachpb.RangeClosedTimestampPolicy with RangeClosedTimestampByPolicyLocality

wenyihu6 committed:
add new proto to sender and receiver
